### ISSUE TITLE
[Proposal] add add basic scaffolding for crud and ability to select fields

### ISSUE THREAD

**Issue by Kasparsu** [laravel/ideas#1310]
I have been using (reluctantly) Yii2 framework while it is mostly hideous work of satan I have come to use and love couple features. Firstly their ActiveRecord (counterpart to laravels Eloquent model) lets you define fields similarly to our fillable but as well as it has extra fields with it it defines what fields will be returned by default and what fields can be queried extra from frontend and that allows for easy API creation that will only send fields requested to the client reducing data size of packet. It also has thing called Active controller that basically adds basic methods get, list, modify, add, delete to be reusable and will work with any model. So you can create model and then controller extend ActiveController fill in a `$model` variable and blam you have basic crud with working actions. There are some methods that can be overriden that allow for more control like manually defining query that gets used and what kinds of sorting, pagination etc can be used. This stuff is pretty easy to implement in Laravel, but it would be cool that this fuctionality could become part of Laravels core system. It could be implemented in such way as it wouldn't break any current functionality or change it but it would allow good addition to basic model view system that we currently enjoy by removing need to copy pasta basically same methods all over again or implement similar logic ourselves

[https://www.yiiframework.com/doc/guide/2.0/en/rest-resources#fields](Expands example)
[https://www.yiiframework.com/doc/guide/2.0/en/rest-controllers#extending-active-controller](ActiveController)

**Comment by mfn** [comment]
> It also has thing called Active controller that basically adds basic methods get, list, modify, add, delete to be reusable

I believe the closest are Eloquents Resources, see https://laravel.com/docs/5.7/eloquent-resources

There are not a single but multiple ideas/approaches mentioned in your idea. I suggest to start a package and pan out how it can/should work.

Finally
> you define fields similarly to our fillable but as well as it has extra fields with it it defines what fields will be returned by default and what fields can be queried extra from frontend and that allows for easy API creation
…
> There are some methods that can be overriden that allow for more control like manually defining query that gets used and what kinds of sorting, pagination etc can be used

Whilst technology-wise orthogonal to what you write (you're talking high-level about REST/CRUD), many of these things are a solved problem with GraphQL too and their resolve concept (see https://github.com/webonyx/graphql-php/ , https://lighthouse-php.netlify.com/ and https://github.com/Folkloreatelier/laravel-graphql for some ideas). You providing a different view here.

**Comment by Kasparsu** [comment]
yes package is a way as well. Eloquent Resources is more of a way to restructure model data rather than actually controlling creating controller actions. GraphQL is something on the lines but it is rather complex system and most projects don't need it.

**Comment by mfn** [comment]
> GraphQL is something on the lines but it is rather complex system and most projects don't need it.

I hear and read that a lot. Having used https://github.com/Folkloreatelier/laravel-graphql I've to say I haven't seen an easier thing to get started _given_ you do understand the GraphQL architecture in general.

But I digress, apologies.
